### ISSUE TITLE
Feature : Automate PyPI Metadata Retrieval for the Compatibility Engine

### ISSUE THREAD

**Issue by vivek0028** [rishabh0510rishabh/EnvForage#269]
# 🚀 Feature Request: Automate PyPI Metadata Retrieval for the Compatibility Engine

## 📌 Overview

Currently, the compatibility engine inside `backend/app/compatibility/matrix/python.py` relies on a hardcoded `PYTHON_MATRIX` dictionary to map ML frameworks (such as PyTorch, TensorFlow, and Ultralytics) to supported Python versions.

There is already a maintainer TODO in the codebase:

```python
# TODO: Automate this from PyPI metadata in a future phase.
```

This feature proposes replacing the static compatibility matrix with a dynamic metadata retrieval system powered by the PyPI JSON API.

---

## 🎯 Goal

Build an automated compatibility engine that:

* Fetches package metadata directly from PyPI
* Parses `Requires-Python` constraints
* Dynamically determines supported Python versions
* Caches responses for performance and rate-limit safety
* Keeps compatibility data automatically up-to-date

---

### Benefits

* Eliminates outdated hardcoded compatibility mappings
* Reduces maintenance overhead
* Improves accuracy for framework version compatibility
* Makes the engine scalable for additional ML libraries
* Enables real-time ecosystem updates

### Resume/Portfolio Impact

This contribution demonstrates:

* API integration skills
* Metadata parsing
* Version constraint handling
* Backend architecture improvements
* Intelligent caching strategies
* Clean Python engineering practices

---

# 🛠 Proposed Implementation

## 1️⃣ Create a PyPI Metadata Fetcher

Use the official PyPI JSON API:

```bash
https://pypi.org/pypi/<package-name>/json
```

Example:

```bash
https://pypi.org/pypi/torch/json
```

Fetch:

* Available versions
* `requires_python`
* Release metadata

---

## 2️⃣ Parse Python Compatibility Constraints

Extract and normalize:

```python
requires_python
```

Examples:

```text
>=3.8
>=3.9,<3.13
~=3.10
```

Use:

```python
packaging.specifiers.SpecifierSet
packaging.version.Version
```

---

## 3️⃣ Generate Dynamic Compatibility Matrix

Convert metadata into a structure like:

```python
{
    "torch": {
        "2.5.0": ["3.9", "3.10", "3.11", "3.12"]
    }
}
```

---

## 4️⃣ Add Smart Caching Layer

Implement:

* In-memory caching
* File-based caching
* Optional Redis support

Suggested TTL:

* 12–24 hours

This avoids:

* Excessive API calls
* Slow startup performance

---

## 5️⃣ Fallback & Error Handling

If PyPI is unavailable:

* Use last cached result
* Gracefully fallback to static defaults

---

# 📂 Suggested File Changes

### New Files

```bash
backend/app/compatibility/services/pypi_client.py
backend/app/compatibility/services/python_metadata_parser.py
backend/app/compatibility/cache/
```

### Updated Files

```bash
backend/app/compatibility/matrix/python.py
```

---

# ✅ Acceptance Criteria

* [ ] Compatibility data fetched dynamically from PyPI
* [ ] `Requires-Python` constraints parsed correctly
* [ ] Framework compatibility generated automatically
* [ ] Caching implemented
* [ ] Graceful fallback handling added
* [ ] Unit tests added
* [ ] Existing API behavior preserved
* [ ] Documentation updated

---

# 🧪 Suggested Packages

```bash
requests
packaging
cachetools
```

---

# 🔥 Bonus Ideas (Optional)

* Support multiple package registries
* Async metadata fetching
* Scheduled background refresh jobs
* Compatibility diff tracking
* Admin endpoint to refresh cache manually

---

# 🏷 Labels

`enhancement` `backend` `python` `api-integration` `good-first-advanced-issue` `NSoC'26` `GSSoC'26` `SSOC'26`

**Comment by rishabh0510rishabh** [comment]
assigned @vivek0369
